### DISCUSSION TITLE
Features for all supported databases available at runtime?

### DISCUSSION THREAD

**Discussion by danielFesenmeyer** [keycloak/keycloak#12116] [Archive]
Hi,

I'm just wondering why the db type has to be specified at build time (see https://www.keycloak.org/server/db), but when starting keycloak it looks like the quarkus features for ALL supported databases are enabled:

```
2022-05-19 11:52:27,798 INFO  [io.quarkus] (main) Installed features: [..., jdbc-h2, jdbc-mariadb, jdbc-mssql, jdbc-mysql, jdbc-oracle, jdbc-postgresql,...]
```

Tested with keycloak 18.0.0. I'm not sure whether this is the expected behavior? And if yes, if there is some configuration option to remove/disable the non-required features?

**Comment by pravsjv** [comment]
I too have a similar question on how we can remove some of the unwanted drivers? and as well, can we configure/add these drivers based on our requirement at runtime instead of loading all of them.

**Comment by missedone** [comment]
have the same question here.
i tried to disable the features with `/opt/keycloak/bin/kc.sh build --features-disabled jdbc-h2,jdbc-mariadb,jdbc-mssql,jdbc-mysql,jdbc-oracle`

but got the error:

```
Invalid value for option '--features-disabled': jdbc-h2,jdbc-mariadb,jdbc-mssql,jdbc-mysql,jdbc-oracle. Expected values are: account-api, account2, account3, admin-api, admin-fine-grained-authz, admin2, authorization, ciba, client-policies, client-secret-rotation, declarative-user-profile, device-flow, docker, dpop, dynamic-scopes, fips, impersonation, js-adapter, kerberos, linkedin-oauth, map-storage, multi-site, par, preview, recovery-codes, scripts, step-up-authentication, token-exchange, transient-users, update-email, web-authn
```

PS: with Keycloak 23.0.6

**Comment by vmuzikar** [comment]
To give you a bit of technical background, the `--db` option actually maps to [`quarkus.datasource.db-kind`](https://quarkus.io/guides/datasource#configuration-reference) Quarkus option which is build time, hence the Keycloak option also needs to be build time.

In the upcoming Keycloak 24, we're actually improving in this area and we're [removing](https://github.com/keycloak/keycloak/pull/25012) the currently unused DB drivers from the classpath, so you shouldn't see them amongst the installed features.
